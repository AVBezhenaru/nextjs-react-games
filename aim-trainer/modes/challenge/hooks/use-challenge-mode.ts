import { useEffect } from 'react';

import { useAppDispatch, useAppSelector } from '../../../../hooks';
import { setModeInfoArray } from '../../../reducers/board-slice';
import { useTimeInfo } from '../../../utils/hooks/use-time-info';
import { useSpawnSpeed } from '../../../utils/hooks/use-spawn-speed';
import { resetGameData, selectGameIsOver, selectGameIsStarted } from '../../../reducers/game-slice';
import { useHitsCounter } from '../../../utils/hooks/use-hits-counter';
import { resetGameStat, selectAccuracy, setStatItems } from '../../../reducers/statistics-slice';
import { DifficultyLevel } from '../../../utils/types/difficulty';

type UseChallengeMode = (diff: DifficultyLevel) => void;

export const useChallengeMode: UseChallengeMode = ({ defaultSpeed, speedIncrementPerSecond }) => {
  const dispatch = useAppDispatch();

  const gameIsStarted = useAppSelector(selectGameIsStarted);
  const gameIsOver = useAppSelector(selectGameIsOver);
  const accuracy = useAppSelector(selectAccuracy);

  const timeFromStartObj = useTimeInfo();
  const targetSpeedObj = useSpawnSpeed(defaultSpeed, speedIncrementPerSecond);
  const hitsObj = useHitsCounter();

  useEffect(() => {
    if (gameIsStarted) {
      dispatch(setModeInfoArray([timeFromStartObj, hitsObj, targetSpeedObj]));
    } else if (gameIsOver) {
      dispatch(
        setStatItems([
          { label: 'Hit Targets', value: hitsObj.value },
          { label: 'Accuracy', value: accuracy },
          { label: 'Final Speed', value: targetSpeedObj.value },
          timeFromStartObj,
        ]),
      );
    }
  }, [timeFromStartObj, hitsObj, targetSpeedObj, gameIsStarted, gameIsOver]);

  useEffect(
    () => () => {
      dispatch(resetGameData());
      dispatch(resetGameStat());
    },
    [],
  );
};
