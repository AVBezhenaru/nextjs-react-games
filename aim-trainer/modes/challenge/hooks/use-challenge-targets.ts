import { useCallback, useEffect, useRef } from 'react';

import { SizeType } from '../../../utils/types/sizes';
import { getRandomNumber } from '../../../utils/lib/get-random-number';
import { useAppDispatch, useAppSelector } from '../../../../hooks';
import { addTarget, selectTargets } from '../../../reducers/targets-slice';
import { selectGameIsStarted } from '../../../reducers/game-slice';
import { selectGameSpeed } from '../../../reducers/statistics-slice';
import { TargetConfiguration } from '../../../utils/types/target';
import { selectBoardSizes } from '../../../reducers/board-slice';
import { selectCurrentGameDifficulty } from '../../../reducers/difficulty-slice';

type GetChallengeTargetConfig = (
  boardSize: SizeType,
  lifetime: number,
  targetSize: number,
) => TargetConfiguration;

const getChallengeTargetConfig: GetChallengeTargetConfig = (boardSize, lifetime, targetSize) => ({
  position: {
    x: getRandomNumber(40, boardSize.w - targetSize - 40),
    y: getRandomNumber(40, boardSize.h - targetSize - 40),
  },
  size: targetSize,
  lifetime,
});

type UseChallengeTargets = () => () => void;

export const useChallengeTargets: UseChallengeTargets = () => {
  const targets = useAppSelector(selectTargets);
  const gameIsStarted = useAppSelector(selectGameIsStarted);
  const speed = useAppSelector(selectGameSpeed);
  const boardSize = useAppSelector(selectBoardSizes);
  const { lifetime, targetSize, maxTargetCount } = useAppSelector(selectCurrentGameDifficulty);

  const currentBoardSize = useRef<SizeType>(boardSize);

  const dispatch = useAppDispatch();

  const createChallengeTarget = useCallback(() => {
    dispatch(addTarget(getChallengeTargetConfig(currentBoardSize.current, lifetime, targetSize)));
  }, [boardSize, lifetime, targetSize]);

  useEffect(() => {
    currentBoardSize.current = boardSize;
  }, [boardSize]);

  useEffect(() => {
    if (gameIsStarted && targets.length < maxTargetCount) {
      if (!targets.length) {
        createChallengeTarget();
      }

      const timerID = setTimeout(() => {
        createChallengeTarget();
      }, 1000 / speed);

      return () => clearTimeout(timerID);
    }
  }, [gameIsStarted, targets]);

  return createChallengeTarget;
};
