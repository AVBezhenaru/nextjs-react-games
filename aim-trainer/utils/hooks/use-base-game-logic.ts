import { useEffect } from 'react';

import { useAppDispatch, useAppSelector } from '../../../hooks';
import {
  resetGameData,
  selectDeathCount,
  selectGameIsStarted,
  selectLives,
  setGameStatus,
} from '../../reducers/game-slice';
import { GameStatus } from '../enums/game-status';
import {
  addTimeFromStart,
  resetGameStat,
  selectGameSpeed,
  setSpeed,
} from '../../reducers/statistics-slice';
import { selectCurrentGameDifficulty, setGameDifficulty } from '../../reducers/difficulty-slice';
import { clearTargets } from '../../reducers/targets-slice';
import { DifficultyLevels } from '../enums/difficulty-levels';

export const useBaseGameLogic = () => {
  const dispatch = useAppDispatch();
  const lives = useAppSelector(selectLives);
  const death = useAppSelector(selectDeathCount);
  const gameIsStarted = useAppSelector(selectGameIsStarted);
  const speed = useAppSelector(selectGameSpeed);
  const { speedIncrementPerSecond } = useAppSelector(selectCurrentGameDifficulty);

  const resetAllGameData = () => {
    dispatch(resetGameData());
    dispatch(resetGameStat());
  };

  useEffect(() => {
    if (lives <= death) {
      dispatch(setGameStatus(GameStatus.Over));
      dispatch(clearTargets());
    }
  }, [lives, death]);

  useEffect(() => {
    dispatch(setGameDifficulty(DifficultyLevels.Normal));

    return resetAllGameData;
  }, []);

  useEffect(() => {
    if (gameIsStarted) {
      const timerID = setInterval(() => {
        dispatch(addTimeFromStart(100));
      }, 100);

      return () => clearInterval(timerID);
    }
  }, [gameIsStarted]);

  useEffect(() => {
    if (gameIsStarted) {
      const timerID = setTimeout(() => {
        console.log(speed, speedIncrementPerSecond);
        dispatch(setSpeed(speed + (speedIncrementPerSecond || 0)));
      }, 1000);

      return () => clearTimeout(timerID);
    }
  }, [gameIsStarted, speed]);
};
