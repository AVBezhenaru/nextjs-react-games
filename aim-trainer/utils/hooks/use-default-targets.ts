import { useCallback, useEffect, useRef } from 'react';
import { useTheme } from 'styled-components';

import { selectBoardSizes } from '../../reducers/board-slice';
import { TargetAnimationTypes, TargetConfiguration, TargetCreator } from '../types/target';
import { selectGameSpeed } from '../../reducers/statistics-slice';
import { addTarget, selectTargets } from '../../reducers/targets-slice';
import { getRandomNumber } from '../lib/get-random-number';
import { useAppDispatch, useAppSelector } from '../../../hooks';
import { selectCurrentGameDifficulty } from '../../reducers/difficulty-slice';
import { selectGameIsStarted } from '../../reducers/game-slice';
import { SizeType } from '../types/sizes';

type GetTargetConfigOptions = {
  boardSize: SizeType;
  lifetime: number;
  targetSize: number;
  animationType: TargetAnimationTypes;
  offset: number;
};
type GetTargetConfig = (options: GetTargetConfigOptions) => TargetConfiguration;

const getTargetConfig: GetTargetConfig = (options) => {
  const { boardSize, lifetime, targetSize, animationType, offset } = options;

  return {
    position: {
      x: getRandomNumber(offset, boardSize.w - targetSize - offset),
      y: getRandomNumber(offset, boardSize.h - targetSize - offset),
    },
    size: targetSize,
    lifetime,
    animationType,
  };
};

type TargetConfig = {
  animationType?: TargetAnimationTypes;
};

type UseDefaultTargets = (config?: TargetConfig) => TargetCreator;

export const useDefaultTargets: UseDefaultTargets = (
  config = { animationType: TargetAnimationTypes.scale },
) => {
  const targets = useAppSelector(selectTargets);
  const gameIsStarted = useAppSelector(selectGameIsStarted);
  const speed = useAppSelector(selectGameSpeed);
  const boardSize = useAppSelector(selectBoardSizes);
  const { lifetime, targetSize, maxTargetCount } = useAppSelector(selectCurrentGameDifficulty);
  const theme = useTheme();

  const currentBoardSize = useRef<SizeType>(boardSize);

  const dispatch = useAppDispatch();

  const createTarget = useCallback(() => {
    dispatch(
      addTarget(
        getTargetConfig({
          boardSize: currentBoardSize.current,
          lifetime,
          targetSize,
          animationType: config.animationType,
          offset: parseInt(theme.borders.baseRadius, 10),
        }),
      ),
    );
  }, [lifetime, targetSize, maxTargetCount]);

  useEffect(() => {
    currentBoardSize.current = boardSize;
  }, [boardSize]);

  useEffect(() => {
    if (gameIsStarted && targets.length < maxTargetCount) {
      if (!targets.length) {
        createTarget();
      }

      const timerID = setTimeout(() => {
        createTarget();
      }, 1000 / speed);

      return () => clearTimeout(timerID);
    }
  }, [gameIsStarted, targets]);

  return createTarget;
};
