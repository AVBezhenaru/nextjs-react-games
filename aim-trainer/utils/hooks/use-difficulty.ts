import { useEffect } from 'react';

import { useAppDispatch, useAppSelector } from '../../../hooks';
import {
  selectCurrentGameDifficulty,
  selectCustomGameDifficulty,
  selectGameDifficulty,
  setCurrentDifficulty,
} from '../../reducers/difficulty-slice';
import { DifficultyLevels } from '../enums/difficulty-levels';
import { resetGameStat, setSpeed } from '../../reducers/statistics-slice';
import { resetGameData, setLives } from '../../reducers/game-slice';
import { DifficultyLevel } from '../types/difficulty';

type UseDifficulty = (levels: {
  easy: DifficultyLevel;
  normal: DifficultyLevel;
  hard: DifficultyLevel;
}) => void;

export const useDifficulty: UseDifficulty = (levels) => {
  const { easy, normal, hard } = levels;

  const dispatch = useAppDispatch();
  const gameDifficultyType = useAppSelector(selectGameDifficulty);
  const customDifficulty = useAppSelector(selectCustomGameDifficulty);
  const currentDifficulty = useAppSelector(selectCurrentGameDifficulty);

  useEffect(() => {
    switch (gameDifficultyType) {
      case DifficultyLevels.Easy:
        dispatch(setCurrentDifficulty(easy));
        break;
      case DifficultyLevels.Normal:
        dispatch(setCurrentDifficulty(normal));
        break;
      case DifficultyLevels.Hard:
        dispatch(setCurrentDifficulty(hard));
        break;
    }
  }, [gameDifficultyType]);

  useEffect(() => {
    if (customDifficulty) {
      dispatch(resetGameStat());
      dispatch(resetGameData());
      dispatch(setCurrentDifficulty(customDifficulty));
    }
  }, [customDifficulty]);

  useEffect(() => {
    dispatch(setLives(currentDifficulty.lives));
    dispatch(setSpeed(currentDifficulty.defaultSpeed));
  }, [currentDifficulty]);
};
