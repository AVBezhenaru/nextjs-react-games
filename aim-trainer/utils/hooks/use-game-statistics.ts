import { useEffect } from 'react';

import { StatCreator } from '../types/statistics';
import { useAppDispatch, useAppSelector } from '../../../hooks';
import { selectGameIsOver } from '../../reducers/game-slice';
import { setStatItems } from '../../reducers/statistics-slice';

type UseGameStatistics = (statCreator: StatCreator) => void;

export const useGameStatistics: UseGameStatistics = (statCreator) => {
  const dispatch = useAppDispatch();
  const gameIsOver = useAppSelector(selectGameIsOver);

  useEffect(() => {
    if (gameIsOver) {
      dispatch(setStatItems(statCreator()));
    }
  }, [gameIsOver]);
};
