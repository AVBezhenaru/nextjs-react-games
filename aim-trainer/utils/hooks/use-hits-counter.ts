import { useMemo } from 'react';

import { ModeInfoObject } from '../types/modes';
import { useAppSelector } from '../../../hooks';
import { selectScore } from '../../reducers/statistics-slice';

type UseHitsCounter = () => ModeInfoObject;

export const useHitsCounter: UseHitsCounter = () => {
  const hits = useAppSelector(selectScore);

  return useMemo<ModeInfoObject>(
    () => ({
      label: 'Hits',
      value: `${hits}`,
    }),
    [hits],
  );
};
