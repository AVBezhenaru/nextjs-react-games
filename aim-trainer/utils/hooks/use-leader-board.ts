import { useEffect, useState } from 'react';

import { StatCreator } from '../types/statistics';
import { useAppDispatch, useAppSelector } from '../../../hooks';
import { selectGameDifficulty } from '../../reducers/difficulty-slice';
import { GameModesPaths } from '../enums/game-modes-paths';
import {
  DifficultyModeData,
  selectLocalModeData,
  setModeData,
} from '../../reducers/leader-list-slice';
import { selectGameIsOver, selectGameIsStarted } from '../../reducers/game-slice';
import { DifficultyLevels } from '../enums/difficulty-levels';

type SortCb = (a: DifficultyModeData, b: DifficultyModeData) => number;

export const useLeaderBoard = (statCreator: StatCreator, mode: GameModesPaths, sortCb: SortCb) => {
  const dispatch = useAppDispatch();

  const currentDifficulty = useAppSelector(selectGameDifficulty);
  const modeData = useAppSelector(selectLocalModeData(mode));
  const gameIsOver = useAppSelector(selectGameIsOver);
  const gameIsStarted = useAppSelector(selectGameIsStarted);

  const [isSetted, setIsSetted] = useState(false);

  useEffect(() => {
    if (gameIsStarted) {
      setIsSetted(false);
    }
  }, [gameIsStarted]);

  useEffect(() => {
    if (gameIsOver && currentDifficulty !== DifficultyLevels.Custom) {
      const localModeData: DifficultyModeData[] = [
        ...(modeData[currentDifficulty] || []),
        {
          statItems: statCreator(),
          timestamp: Date.now(),
        },
      ];

      localModeData.sort(sortCb);

      if (!isSetted) {
        dispatch(
          setModeData({
            difficulty: currentDifficulty,
            data: localModeData.slice(0, 10),
            mode,
          }),
        );

        setIsSetted(true);
      }
    }
  }, [gameIsOver]);
};
