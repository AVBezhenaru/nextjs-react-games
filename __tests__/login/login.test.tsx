/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react';

import LoginForm from '../../user/forms/loginForm';
import '@testing-library/jest-dom';
import { useAppSelector } from '../../hooks';

jest.mock('react-redux');
(useAppSelector as jest.MockedFunction<any>).mockReturnValue([]);

describe('Login form test', () => {
  it('Render test', () => {
    render(<LoginForm />);
    expect(screen.getByTestId('login-form')).toBeInTheDocument();
  });
  it('Snapshot login form', () => {
    const component = render(<LoginForm />);
    expect(component).toMatchSnapshot();
  });
});
