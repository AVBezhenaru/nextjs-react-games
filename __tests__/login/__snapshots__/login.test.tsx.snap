// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form test Snapshot login form 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="sc-bcXHqe jqemYE"
      >
        <form
          class="sc-jSUZER hOSZEZ"
          data-testid="login-form"
        >
          <div
            class="sc-gswNZR iLSRkP"
          >
            <a
              href="https://www.google.com"
            >
              <img
                alt="google"
                class="sc-dkrFOg fgudp"
              />
            </a>
            <a
              href="https://www.facebook.com"
            >
              <img
                alt="facebook"
                class="sc-dkrFOg fgudp"
              />
            </a>
            <a
              href="https://www.twitter.com"
            >
              <img
                alt="twitter"
                class="sc-dkrFOg fgudp"
              />
            </a>
          </div>
          <h1
            class="sc-eDvSVe kUwkqi"
          >
            <span
              class="sc-hLBbgP iQYZFM"
            >
              or with Email
            </span>
          </h1>
          <input
            class="sc-gKPRtg ldvNAE"
            name="username"
            placeholder="User Name  or  email..."
          />
          <input
            class="sc-gKPRtg ldvNAE"
            name="password"
            placeholder="Password"
          />
          <button
            class="sc-iBYQkv hYjtfu"
            type="submit"
          >
            Register
          </button>
          <p
            class="sc-ftTHYK kQBTUj"
          >
            Don\`t have an account?
            <a
              href="/registration"
              style="color: rgb(244, 97, 25);"
            >
              Sign Up
            </a>
          </p>
        </form>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="sc-bcXHqe jqemYE"
    >
      <form
        class="sc-jSUZER hOSZEZ"
        data-testid="login-form"
      >
        <div
          class="sc-gswNZR iLSRkP"
        >
          <a
            href="https://www.google.com"
          >
            <img
              alt="google"
              class="sc-dkrFOg fgudp"
            />
          </a>
          <a
            href="https://www.facebook.com"
          >
            <img
              alt="facebook"
              class="sc-dkrFOg fgudp"
            />
          </a>
          <a
            href="https://www.twitter.com"
          >
            <img
              alt="twitter"
              class="sc-dkrFOg fgudp"
            />
          </a>
        </div>
        <h1
          class="sc-eDvSVe kUwkqi"
        >
          <span
            class="sc-hLBbgP iQYZFM"
          >
            or with Email
          </span>
        </h1>
        <input
          class="sc-gKPRtg ldvNAE"
          name="username"
          placeholder="User Name  or  email..."
        />
        <input
          class="sc-gKPRtg ldvNAE"
          name="password"
          placeholder="Password"
        />
        <button
          class="sc-iBYQkv hYjtfu"
          type="submit"
        >
          Register
        </button>
        <p
          class="sc-ftTHYK kQBTUj"
        >
          Don\`t have an account?
          <a
            href="/registration"
            style="color: rgb(244, 97, 25);"
          >
            Sign Up
          </a>
        </p>
      </form>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
